namecache/namecache.go

```diff
@@ -0,0 +1,207 @@
+		dag:     dag,
+		follows: make(map[string]func()),
+	}
+}
+
+// Follow spawns a goroutine that periodically resolves a name
+// and (when dopin is true) pins it in the background
+func (nc *nameCache) Follow(name string, prefetch bool, followInterval time.Duration) error {
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	if !strings.HasPrefix(name, "/ipns/") {
+		name = "/ipns/" + name
+	}
+
+	if _, ok := nc.follows[name]; ok {
+		return fmt.Errorf("already following %s", name)
+	}
+
+	ctx, cancel := context.WithCancel(nc.ctx)
+	go nc.followName(ctx, name, prefetch, followInterval)
+	nc.follows[name] = cancel
+
+	return nil
+}
+
+// Unfollow cancels a follow
+func (nc *nameCache) Unfollow(name string) error {
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	if !strings.HasPrefix(name, "/ipns/") {
+		name = "/ipns/" + name
+	}
+
+	cancel, ok := nc.follows[name]
+	if !ok {
+		return fmt.Errorf("unknown name %s", name)
+	}
+
+	cancel()
+	delete(nc.follows, name)
+	return nil
+}
+
+// ListFollows returns a list of names currently being followed
+func (nc *nameCache) ListFollows() []string {
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	follows := make([]string, 0, len(nc.follows))
+	for name := range nc.follows {
+		follows = append(follows, name)
+	}
+
+	return follows
+}
+
+func (nc *nameCache) followName(ctx context.Context, name string, prefetch bool, followInterval time.Duration) {
+	emptynode := new(dag.ProtoNode)
+
+	c, err := nc.resolveAndUpdate(ctx, name, prefetch, emptynode.Cid())
+	if err != nil {
+		log.Errorf("Error following %s: %s", name, err.Error())
+	}
+
+	ticker := time.NewTicker(followInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			c, err = nc.resolveAndUpdate(ctx, name, prefetch, c)
+
+			if err != nil {
+				log.Errorf("Error following %s: %s", name, err.Error())
+			}
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (nc *nameCache) resolveAndUpdate(ctx context.Context, name string, prefetch bool, oldcid cid.Cid) (cid.Cid, error) {
+	ptr, err := nc.resolve(ctx, name)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	newcid, err := pathToCid(ptr)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	if newcid.Equals(oldcid) || !prefetch {
+		return newcid, nil
+	}
+
+	oldnd, err := nc.dag.Get(ctx, oldcid)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	newnd, err := nc.dag.Get(ctx, newcid)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	changes, err := dagutils.Diff(ctx, nc.dag, oldnd, newnd)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	log.Debugf("fetching changes in %s (%s -> %s)", name, oldcid, newcid)
+	for _, change := range changes {
+		if change.Type == iface.DiffRemove {
+			continue
+		}
+
+		toFetch, err := nc.dag.Get(ctx, change.After)
+		if err != nil {
+			return cid.Undef, err
+		}
+
+		// just iterate over all nodes
+		walker := ipld.NewWalker(ctx, ipld.NewNavigableIPLDNode(toFetch, nc.dag))
+		if err := walker.Iterate(func(node ipld.NavigableNode) error {
+			return nil
```

Comment:
It would be nice to add cids of nodes we iterate to some set and skip children we have already seen, but I'm not sure what's the best way to do this.

cc @schomatis

Reply:
> It would be nice to add cids of nodes we iterate to some set and skip children we have already seen, but I'm not sure what's the best way to do this.

1. Solution at the `Visitor` level.

The simplest way would be for the caller to keep that set and use it in the `Visitor`, in there we could check if the CID of the current child the `Walker` is pointing to has already been visited and if so, skip it, something like:

```Go

// For every child of the current `node`:
    childCid := node.Links()[walker.ActiveChildIndex()]
    if _, ok := visited[childCid]; ok {
        walker.NextChild();
        // Skip this child, already visited.
    } else {
        break; // out of the `for`, so when this call to `Visit` ends the `Walker`
        // will iterate into this child we haven't visited.
    }

// If we have skipped all of the children the `Walker` will just continue
// the iteration going up to the current `node`'s parent.
```

The downside of this approach is that `NavigableIPLDNode` still prefetches many children in advance, so if we didn't skip all of the children, if we request at least one, `NavigableIPLDNode` will also start prefetching many more after that one (that we may end up actually skipping).

2. Solution at the `NavigableIPLDNode` level.

In that case `NavigableIPLDNode` should be extended to handle this scenario (which would actually be a very nice feature). (Note here that `Walker` has no idea what an IPLD node is, it only operates on the very abstract notion of a DAG of any kind of node that implements the `NavigableNode` interface; the motivation behind this decoupling was to allow the possibility of using the `Walker` for other types of DAG like the dependency graphs that Gx iterates over, although at the moment IPLD is the only type of node that the `Walker` knows).

The way I think this could be implemented (but this is just an untested idea) would be to modify the [`preload`](https://github.com/ipfs/go-ipld-format/blob/ba07363e1f1f53c7fe828012a80e0ed7ae974d1a/navipld.go#L89-L96) function to, instead of prefetching a contiguous batch of `preloadSize` nodes through `GetNodes`, iterate the `childCIDs` one at a time and skip the ones already visited, calling each prefetch separately with `GetNode` (singular).

How should the map of visited nodes be handled in that case? Every `NavigableIPLDNode` is independent of its parent and children, there's no central authority, the way they communicate with each other is through the `context.Context` that was used in `NewWalker`, that context will be passed to every `FetchChild` issued by the Walker when going down the DAG. (Maybe this is too coupled and the `NavigableIPLDNode` should be extended to have its own context now, that gets passed from parent to child.) So the map could be stored there, and there could be a check that if there's no map of visited nodes, then just prefetch everything as `preload` does now. At the moment the `Walker` doesn't visit the DAG in parallel so this map would be safe (but it should be documented that in future refactorings it may need to be protected from multiple accesses).

/cc @hsanjuan who is interested in something very similar, so maybe you two could join forces here.

Reply:
(@hsanjuan in your case, inside the `Visitor` you could artificially add the CIDs of the nodes you want to skip entirely in this map, which would result in the same effect.)